### PR TITLE
Rollout management issues

### PR DESCRIPTION
**Pull request by venu1278**
fixes for Rollout management issues.
i)While  create new Rollout, RolloutName is not clickable unless status is to change from 'CREATE' to 'READY ' state.
ii)If total number of target groups is zero,then  displaying empty.
iii)The style of Action menu is now same as DeploymentView right click option of target table.
iV)The 'Finished %' is dynamically updates in RolloutGroupList while Rollout is started.

Signed-off-by: venugopal boodidadinne  venugopal.boodidadinne@in.bosch.com

### CODE REVIEW

**Comment by kaizimmerm** [comment]
Please describe in the pull request what you have fixed. Remove Bosch internal Jira numbers.

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/renderers/LinkRenderer.java [hunk 1]
```diff
@@ -21,11 +24,22 @@
...
+        final String targetStatus = nameStatusMap.get("status");
+        button.setText(targetName);
         applystyle(button);
         // this is to allow the button to disappear, if the text is null
-        button.setVisible(text != null);
-        button.getElement().setId(new StringBuilder("link").append(".").append(text).toString());
+        button.setVisible(targetName != null);
+        button.getElement().setId(new StringBuilder("link").append(".").append(targetName).toString());
+
+        if (null != targetStatus && targetStatus.equalsIgnoreCase("CREATING")) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/renderers/LinkRenderer.java]: swtich null parameter

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/renderers/LinkRenderer.java [hunk 2]
```diff
@@ -21,11 +24,22 @@
...
+            button.getElement().setAttribute("enabled", "false");
+        } else {
+            button.getElement().setAttribute("enabled", "true");
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/renderers/LinkRenderer.java]: set enabled attribut direct.
button.getElement().setAttribute("enabled", theIfStatment);

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java [hunk 3]
```diff
@@ -327,10 +336,34 @@ public String getStyle(final CellReference cellReference) {
         });
     }
 
+    private void addGeneratedProperties() {
+        GeneratedPropertyContainer gpContainer = (GeneratedPropertyContainer) getContainerDataSource();
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java]: Use meaningful name not gpContainer

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java [hunk 4]
```diff
@@ -327,10 +336,34 @@ public String getStyle(final CellReference cellReference) {
...
+        gpContainer.addGeneratedProperty(SPUILabelDefinitions.VAR_NAME, new PropertyValueGenerator<String>() {
+            private static final long serialVersionUID = -9203261132281441831L;
+
+            @Override
+            public String getValue(Item item, Object itemId, Object propertyId) {
+                String statusVal = (String) item.getItemProperty(SPUILabelDefinitions.VAR_STATUS).getValue().toString();
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java]: Use meaningful name not statusVal

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java [hunk 5]
```diff
@@ -338,6 +371,22 @@ private void onClickOfRolloutName(RendererClickEvent event) {
         eventBus.publish(this, RolloutEvent.SHOW_ROLLOUT_GROUPS);
     }
 
+    private String getRolloutName(final String text) {
+        String nameSequence = null;
+        String name = null;
+        if (null != text) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java]: 1. early return
2. switch null paramter

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java [hunk 6]
```diff
@@ -338,6 +371,22 @@ private void onClickOfRolloutName(RendererClickEvent event) {
         eventBus.publish(this, RolloutEvent.SHOW_ROLLOUT_GROUPS);
     }
 
+    private String getRolloutName(final String text) {
+        String nameSequence = null;
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java]: Initialize with null is not necessary see sonar rules

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java [hunk 7]
```diff
@@ -163,6 +180,28 @@ protected void addContainerProperties() {
...
+    private void addGeneratedProperties() {
+        GeneratedPropertyContainer gpContainer = (GeneratedPropertyContainer) getContainerDataSource();
+        gpContainer.addGeneratedProperty(SPUILabelDefinitions.VAR_NAME, new PropertyValueGenerator<String>() {
+            private static final long serialVersionUID = -9203261132281441831L;
+
+            @Override
+            public String getValue(Item item, Object itemId, Object propertyId) {
+                String name = (String) item.getItemProperty(SPUILabelDefinitions.VAR_NAME).getValue().toString();
+                Map<String, String> nameStatusMap = new HashMap<>();
+                nameStatusMap.put("name", name);
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java]: is it better to usw constant for "name". You used it several times

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java [hunk 8]
```diff
@@ -304,13 +343,22 @@ private String getDescription(CellReference cell) {
...
-            return cell.getProperty().getValue().toString();
+            return  getNameToolTip(cell.getProperty().getValue().toString());
         } else if (SPUILabelDefinitions.VAR_TOTAL_TARGETS_COUNT_STATUS.equals(cell.getPropertyId())) {
             return DistributionBarHelper
                     .getTooltip(((TotalTargetCountStatus) cell.getValue()).getStatusTotalCountMap());
         }
         return null;
     }
+    
+    private String getNameToolTip(final String text) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java]: Duplicate code to RolloutListGrid line 527 to 530

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java [hunk 9]
```diff
@@ -304,13 +343,22 @@ private String getDescription(CellReference cell) {
...
+        String name = null;
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java]: Not initialize with null

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/utils/HawkbitCommonUtil.java [hunk 10]
```diff
@@ -1369,6 +1369,24 @@ public static String getFormattedString(Map<Status, Long> details) {
...
+    /**
+     * Get the formatted string of status and target name.
+     * @param details
+     *          details of status and name
+     * @return
+     */
+    public static String getNameStatusFormattedString(Map<String, String> details) {
+        StringBuilder val = new StringBuilder();
+        String finalVal = null;
+        if (null != details && !details.isEmpty()) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/utils/HawkbitCommonUtil.java]: 1. early return
2. switch null parameter

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/renderers/LinkRenderer.java [hunk 11]
```diff
@@ -21,11 +24,21 @@
...
-        button.setVisible(text != null);
-        button.getElement().setId(new StringBuilder("link").append(".").append(text).toString());
+        button.setVisible(targetName != null);
+        button.getElement().setId(new StringBuilder("link").append(".").append(targetName).toString());
+        /*
+         * checking target Status for applying button style. If target status is
+         * not "CREATING", then the Rollout button is applying hyperlink style
+         */
+        button.getElement().setAttribute("enabled",
+                targetStatus != null && "CREATING".equalsIgnoreCase(targetStatus) ? "false" : "true");
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/renderers/LinkRenderer.java]: use a boolean paramter for the if statement so you can use it twice.

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java [hunk 12]
```diff
@@ -338,6 +374,26 @@ private void onClickOfRolloutName(RendererClickEvent event) {
...
+        final String[] nameArray = getReformatData(text);
+        if (nameArray != null && nameArray.length > 0) {
+            return nameArray[1];
+        }
+        return "";
+    }
+
+    private String[] getReformatData(final String text) {
+        String nameSequence;
+        if (text != null) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java]: Early return if (text == null) { return null}

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java [hunk 13]
```diff
@@ -338,6 +374,26 @@ private void onClickOfRolloutName(RendererClickEvent event) {
...
+            final String[] strArray = text.split(",");
+            if (strArray.length > 0) {
+                nameSequence = strArray[0];
+                return nameSequence.split(":");
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java]: reduce code: retrun strArray[0].split()... So you need no local paramter

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/utils/HawkbitCommonUtil.java [hunk 14]
```diff
@@ -1371,6 +1371,24 @@ public static String getFormattedString(Map<Status, Long> details) {
     }
 
     /**
+     * Get the formatted string of status and target name.
+     * 
+     * @param details
+     *            details of status and name
+     * @return
+     */
+    public static String getNameStatusFormattedString(Map<String, String> details) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/utils/HawkbitCommonUtil.java]: Use getFormattedString methode. This methide is duplicated code

Comment (venu1278) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/utils/HawkbitCommonUtil.java]: getFormattedString method is deleted as it is not used any where.

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/utils/HawkbitCommonUtil.java [hunk 15]
```diff
@@ -1351,23 +1351,23 @@ private static boolean isNoTargets(final Long... statusCount) {
...
      */
-    public static String getFormattedString(Map<Status, Long> details) {
+    public static String getNameStatusFormattedString(Map<String, String> details) {
         StringBuilder val = new StringBuilder();
-        if (details == null || details.isEmpty()) {
-            return null;
-        }
-        for (Entry<Status, Long> entry : details.entrySet()) {
-            val.append(entry.getKey()).append(":").append(entry.getValue()).append(",");
+        if (details != null && !details.isEmpty()) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/utils/HawkbitCommonUtil.java]: Use EARLY return. if (details == null || details.isEmpty()) --->return

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java [hunk 16]
```diff
@@ -163,6 +179,28 @@ protected void addContainerProperties() {
                 false);
         rolloutGroupGridContainer.addContainerProperty(SPUILabelDefinitions.VAR_TOTAL_TARGETS_COUNT_STATUS,
                 TotalTargetCountStatus.class, null, false, false);
+        addGeneratedProperties();
+    }
+    
+    private void addGeneratedProperties() {
+        GeneratedPropertyContainer gpContainer = (GeneratedPropertyContainer) getContainerDataSource();
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java]: gpContainer is not a good name

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/utils/HawkbitCommonUtil.java (hunk 15 above)
Comment (venu1278) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/utils/HawkbitCommonUtil.java]: getFormattedString method id removed as it is not using any where .

Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/utils/HawkbitCommonUtil.java]: Yes that's okay but use early return for the if statement.
see http://programmers.stackexchange.com/questions/18454/should-i-return-from-a-function-early-or-use-an-if-statement

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/renderers/LinkRenderer.java [hunk 17]
```diff
@@ -21,11 +24,21 @@
...
         // this is to allow the button to disappear, if the text is null
-        button.setVisible(text != null);
-        button.getElement().setId(new StringBuilder("link").append(".").append(text).toString());
+        button.setVisible(targetName != null);
+        button.getElement().setId(new StringBuilder("link").append(".").append(targetName).toString());
+        /*
+         * checking target Status for applying button style. If target status is
+         * not "CREATING", then the Rollout button is applying hyperlink style
+         */
+        final boolean isStatusCreate = targetStatus != null && "CREATING".equalsIgnoreCase(targetStatus);
```
Comment (michahirsch) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/renderers/LinkRenderer.java]: For `CREATING` a constant or even better the enum which defines the status can be used.

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/renderers/LinkRenderer.java [hunk 18]
```diff
@@ -21,11 +24,21 @@
 public class LinkRenderer extends ButtonRenderer {
     @Override
     public void render(RendererCellReference cell, String text, Button button) {
-        button.setText(text);
+        Map<String, String> nameStatusMap = formatData(text);
```
Comment (michahirsch) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/renderers/LinkRenderer.java]: I'm not sure what is in the `text` value to split it up into a map to then retrieve the `name` and the `status` is this Vaadin related? Can this maybe documented or is this documented?

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/renderers/LinkRenderer.java [hunk 19]
```diff
@@ -33,10 +46,19 @@ private void applystyle(Button button) {
...
+
     private String getStyle(final String style) {
         return new StringBuilder(style).append(" ").append(VButton.CLASSNAME).append("-").append(style).toString();
     }
+
+    private Map<String, String> formatData(String input) {
+        Map<String, String> details = new HashMap<>();
+        String[] tempData = input.split(",");
+        for (String targetData : tempData) {
+            String[] targetDataList = targetData.split(":");
```
Comment (michahirsch) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/renderers/LinkRenderer.java]: Is this defintely always an array of two elements? Because we hard accessing by index zero and one, what if in the value is no colon to split? Then there will no index one, maybe we should ensure it?

Comment (venu1278) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/renderers/LinkRenderer.java]: Here the input parameter comes like "name:rolloutName,status:rolloutStatus". This formation is done by addGeneratedProperties() in RolloutListGrid.java.  The reason for appending status with name is applying the  link style for RolloutName button based on Rollout Status in LinkRenderer.
